lib/basedriver/commands/execute-child.js

```diff
@@ -0,0 +1,128 @@
+import _ from 'lodash';
+import B from 'bluebird';
+import vm from 'vm';
+import log from '../logger';
+import { attach } from 'webdriverio';
+
+// duplicate defining these keys here so we don't need to re-load a huge appium
+// dependency tree into memory just to run a wdio script
+const W3C_ELEMENT_KEY = 'element-6066-11e4-a52e-4f735466cecf';
+const MJSONWP_ELEMENT_KEY = 'ELEMENT';
+
+async function runScript (driverOpts, script, timeout) {
+  // set up fake logger
+  const logLevels = ['error', 'warn', 'log'];
+  const logs = {};
+  const consoleFns = {};
+  for (const level of logLevels) {
+    logs[level] = [];
+    consoleFns[level] = (...logMsgs) => logs[level].push(...logMsgs);
+  }
+
+  const driver = attach(driverOpts);
+
+  const fullScript = buildScript(script);
+  // the timeout here will not matter really, but set it anyway to be on the
+  // safe side
+  const vmCtx = vm.runInNewContext(fullScript, {}, {timeout});
+
+  // run the driver script, giving user access to the driver object, a fake
+  // console logger, and a promise library
+  log.info('Running driver script in Node vm');
+  let result = await vmCtx(driver, consoleFns, B);
+
+  log.info('Ensuring driver script result is appropriate type for return');
+  result = coerceScriptResult(result);
+  return {result, logs};
+}
+
+/**
+ * Embed a user-generated script inside a method which takes only the
+ * predetermined objects we specify
+ *
+ * @param {string} script - the javascript to execute
+ *
+ * @return {string} - the full script to execute
+ */
+function buildScript (script) {
+  return `(async function execute (driver, console, Promise) {
```

Comment:
aren't console and Promise shadowing standard object names?

Reply:
these standard object names are not available in the VM. it's literally just a JS language interpreter, it has no standard web or node objects available, so we have to provide everything.